File: 206. Reverse Linked List.cpp
206. Reverse Linked List
Reverse a singly linked list.

click to show more hints.

Subscribe to see which companies asked this question

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        stack<int> s;
        ListNode* p = head;
        while(p != NULL) {
            s.push(p->val);
            p = p->next;
        }
        p = head;
        while(p != NULL) {
            p->val = s.top();
            s.pop();
            p = p->next;
        }
        return head;
    }
};

File: Swift/206. Reverse Linked List.swift
// 206. Reverse Linked List
// 20 ms, 63.85%
func reverseList(_ head: ListNode?) -> ListNode? {
    var stack = [Int](), p = head
    while let t = p {
        stack.append(t.val)
        p = t.next
    }
    p = head
    while let t = p {
        t.val = stack.popLast()!
        p = t.next
    }
    return head
}